otter/convergence/selfheal.py

```diff
@@ -0,0 +1,169 @@
+"""
+Self heal service
+"""
+
+from effect import Effect
+from effect.do import do, do_return
+
+from kazoo.exceptions import LockTimeout
+from kazoo.protocol.states import KazooState
+
+from toolz.curried import filter
+
+from twisted.application.internet import TimerService
+from twisted.application.service import MultiService
+from twisted.internet.defer import inlineCallbacks, returnValue
+
+from txeffect import perform
+
+from otter.convergence.composition import tenant_is_enabled
+from otter.convergence.service import trigger_convergence
+from otter.log.intents import with_log
+from otter.models.intents import GetAllValidGroups, GetScalingGroupInfo
+from otter.models.interface import ScalingGroupStatus
+from otter.util.zk import GetChildren
+
+
+class SelfHeal(MultiService, object):
+    """
+    A service that triggers convergence on all the groups on interval basis.
+    Only one node is allowed to do this.
+    """
+
+    def __init__(self, dispatcher, kz_client, interval, log, clock,
+                 config_func):
+        super(SelfHeal, self).__init__()
+        self.disp = dispatcher
+        self.log = log.bind(otter_service="selfheal")
+        self.lock = kz_client.Lock("/selfheallock")
+        self.kz_client = kz_client
+        self.clock = clock
+        self.time_range = interval - 5
+        self.config_func = config_func
+        self.calls = []
+        timer = TimerService(
+            interval, lambda: self._converge_all().addErrback(self.log.err))
+        timer.clock = clock
+        timer.setServiceParent(self)
+
+    def stopService(self):
+        """
+        Stop service by cancelling any remaining scheduled calls and releasing
+        the lock
+        """
+        super(SelfHeal, self).stopService()
+        self._cancel_scheduled_calls()
+        return self.lock.release()
+
+    def _cancel_scheduled_calls(self):
+        """
+        Cancel any remaining scheduled calls
+        """
+        active = 0
+        for call in self.calls:
+            if call.active():
+                active += 1
+                call.cancel()
+        self.calls = []
+        return active
+
+    def health_check(self):
+        """
+        Return about whether this object has lock
+        """
+        d = is_lock_acquired(self.disp, self.lock)
+        return d.addCallback(lambda b: (True, {"has_lock": b}))
+
+    def _perform(self):
+        d = perform(self.disp, get_groups_to_converge(self.config_func))
+        d.addCallback(self._setup_converges)
```

Comment:
non-blocking, but:

I have some general feedback about the algorithm you're using to do the timed work here.

It involves managing a lot of state, so it's somewhat error prone.  I don't spot any _actual_ errors but I suspect it may be a bit tricky to maintain.

Rather than having one looping call which sets up another set of looping calls, which, if the network slows down, might fail to apply backpressure and therefore end up backing up, I think you should switch to a service which has only a single `LoopingCall` that just figures out what the "next thing to do" is.

For example, initially, the first thing to do is to get the groups to converge.  Then you can return that `Deferred` from your `LoopingCall` so you won't be called again until it's done.  Once you've got the groups, then the next "thing to do" is to converge all the groups.  This batch-convergence should probably be wrapped in a `gatherResults`.  When you run out of groups and they're all converged, the next thing to do is to get groups to converge again.  This little state machine can completely eliminate the management of outstanding calls, since you'd only have one outstanding delayed call at a time.  This would involve having a much higher frequency.

Does this make sense as a follow-up change?  ideally most of the tests would be similar...